presto-orc/src/main/java/io/prestosql/orc/reader/UnionColumnReader.java

```diff
@@ -0,0 +1,318 @@
+        }
+
+        Block[] blocks = new Block[fieldReaders.size() + 1];
+
+        byte[] tags = dataStream.next(positionCount);
+        blocks[0] = new ByteArrayBlock(positionCount, Optional.empty(), tags);
+
+        boolean[][] valueIsNonNull = new boolean[fieldReaders.size()][positionCount];
+        int[] nonNullValueCount = new int[fieldReaders.size()];
+        for (int i = 0; i < positionCount; i++) {
+            valueIsNonNull[tags[i]][i] = true;
+            nonNullValueCount[tags[i]]++;
+        }
+
+        for (int i = 0; i < fieldReaders.size(); i++) {
+            Type fieldType = type.getTypeParameters().get(i + 1);
+            if (nonNullValueCount[i] > 0) {
+                ColumnReader reader = fieldReaders.get(i);
+                reader.prepareNextRead(nonNullValueCount[i]);
+                Block rawBlock = blockFactory.createBlock(nonNullValueCount[i], reader::readBlock, true);
+                blocks[i + 1] = new LazyBlock(positionCount, new UnpackLazyBlockLoader(rawBlock, fieldType, valueIsNonNull[i]));
+            }
+            else {
+                blocks[i + 1] = new RunLengthEncodedBlock(
+                        fieldType.createBlockBuilder(null, 1).appendNull().build(),
+                        positionCount);
+            }
+        }
+        return blocks;
+    }
+
+    @Override
+    public void close()
+    {
+        try (Closer closer = Closer.create()) {
+            for (ColumnReader structField : fieldReaders) {
+                closer.register(structField::close);
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        long retainedSizeInBytes = INSTANCE_SIZE;
+        for (ColumnReader structField : fieldReaders) {
+            retainedSizeInBytes += structField.getRetainedSizeInBytes();
+        }
+        return retainedSizeInBytes;
+    }
+
+    private static final class UnpackLazyBlockLoader
+            implements LazyBlockLoader
+    {
+        private final Block denseBlock;
+        private final Type type;
+        private final boolean[] valueIsNonNull;
+
+        public UnpackLazyBlockLoader(Block denseBlock, Type type, boolean[] valueIsNonNull)
+        {
+            this.denseBlock = requireNonNull(denseBlock, "denseBlock is null");
+            this.type = requireNonNull(type, "type is null");
+            this.valueIsNonNull = requireNonNull(valueIsNonNull, "valueIsNonNull");
+        }
+
+        @Override
+        public Block load()
+        {
+            Block loadedDenseBlock = denseBlock.getLoadedBlock();
+            BlockBuilder unpackedBlock = type.createBlockBuilder(null, valueIsNonNull.length);
+
+            int denseBlockPosition = 0;
+            for (boolean isNonNull : valueIsNonNull) {
+                if (isNonNull) {
+                    type.appendTo(loadedDenseBlock, denseBlockPosition++, unpackedBlock);
```

Comment:
Instead of copying the data use a dictionary block.

Reply:
@dain Thanks for the pointer, but I didn't find a good way to handle `NULL`s. Is there a way to append a `NULL` to the raw block so that it can serve as a dictionary?

Reply:
Ah, yes.  This is the same problem we had with unnest.  In that case, we scanned the block for a null and if present, we used that; otherwise we copied.  We can leave this for now.